### PR TITLE
[Feature] Support select binlog

### PR DESCRIPTION
**Pull request by banmoy**
## What type of PR is this：
- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Add binlog source operator, and support select clause for binlog. Currently the operator return the mocked data, and will return the real binlog data in another PR. The select clause need to set session variable `enable_mv_planner` to `true`, and use `[_BINLOG_]` hint to indicate to read from binlog. Note that select clause is mainly used for testing currently, and it's not user-oriented.
```
CREATE TABLE `test_binlog` (
  `c0` int(11) NULL COMMENT "",
  `c1` int(11) NULL COMMENT "",
  `c2` int(11) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`c0`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`c0`) BUCKETS 2
PROPERTIES (
"replication_num" = "1"
);

set enable_mv_planner=true;
select * from test_binlog [_BINLOG_];

+------+------+------+------------+-----------------+----------------+-------------------+
| c0   | c1   | c2   | _binlog_op | _binlog_version | _binlog_seq_id | _binlog_timestamp |
+------+------+------+------------+-----------------+----------------+-------------------+
|    0 |    1 |    2 |          3 |               4 |              5 |                 6 |
|    1 |    2 |    3 |          4 |               5 |              6 |                 7 |
|    2 |    3 |    4 |          5 |               6 |              7 |                 8 |
|    3 |    4 |    5 |          6 |               7 |              8 |                 9 |
|    4 |    5 |    6 |          7 |               8 |              9 |                10 |
|    5 |    6 |    7 |          8 |               9 |             10 |                11 |
|    6 |    7 |    8 |          9 |              10 |             11 |                12 |
+------+------+------+------------+-----------------+----------------+-------------------+
```

## Checklist:

- [X] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 17 / 28 (60.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|1|3|33.33%|[118, 119]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|5|40.00%|[2857, 2858, 2859]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|6|12|50.00%|[295, 296, 297, 298, 299, 300]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/BinlogScanNode.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 17 / 28 (60.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|1|3|33.33%|[118, 119]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|5|40.00%|[2857, 2858, 2859]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|6|12|50.00%|[295, 296, 297, 298, 299, 300]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/BinlogScanNode.java|6|6|100.00%|[]|

**Comment by banmoy** [comment]
run starrocks_admit_test

File: be/src/connector/binlog_connector.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+Status BinlogDataSource::_mock_chunk(Chunk* chunk) {
+    int32_t num = _chunk_num.fetch_add(1, std::memory_order_relaxed);
+    if (num >= 2) {
+        return Status::EndOfFile(fmt::format("Has sent {} chunks", num));
+    }
+
+    for (int row = 0; row < 10; row++) {
+        for (int col = 0; col < chunk->num_columns(); col++) {
+            LogicalType type = _binlog_read_schema.field(col)->type()->type();
```
Comment (caneGuy) [inline be/src/connector/binlog_connector.cpp:147]: we can not 'select a from xxx _BINGLOG' if a is a data column which is varchar type, am i right?

Comment (banmoy) [reply be/src/connector/binlog_connector.cpp:147]: Yes, you are right. This PR only mock data for some types, and it's enough for testing. Of course we can mock data for more types if needed.

File: be/src/exec/pipeline/scan/stream_scan_operator.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#include "exec/workgroup/work_group_fwd.h"
+#include "storage/chunk_helper.h"
+
+namespace starrocks {
+
+class ScanNode;
+
+namespace pipeline {
+
+class StreamScanOperatorFactory final : public ConnectorScanOperatorFactory {
```
[the remark this answers is not in the document]
Comment (banmoy) [reply be/src/exec/pipeline/scan/stream_scan_operator.h]: Done

File: be/src/connector/binlog_connector.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    return Status::OK();
+}
+
+void BinlogDataSource::close(RuntimeState* state) {}
+
+Status BinlogDataSource::get_next(RuntimeState* state, ChunkPtr* chunk) {
+    SCOPED_RAW_TIMER(&_cpu_time_ns);
+    _init_chunk(chunk, state->chunk_size());
+    // TODO replace with BinlogReader
+    return _mock_chunk(chunk->get());
```
[the remark this answers is not in the document]
Comment (banmoy) [reply be/src/connector/binlog_connector.cpp:60]: Yes, will support it in another pr

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 43 / 46 (93.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|17|82.35%|[2857, 2858, 2859]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/BinlogScanNode.java|16|16|100.00%|[]|

File: fe/fe-core/src/test/java/com/starrocks/utframe/UtFrameUtils.java [hunk 4]
```diff
@@ -381,8 +381,12 @@ public static Pair<CreateMaterializedViewStatement, ExecPlan> planMVMaintenance(
...
     }
 
-    public static Pair<String, ExecPlan> getPlanAndFragment(ConnectContext connectContext, String originStmt)
-            throws Exception {
+    public static Pair<String, ExecPlan> getPlanAndFragment(ConnectContext connectContext, String originStmt) throws Exception {
+        return getPlanAndFragment(connectContext, originStmt, true);
+    }
+
+    public static Pair<String, ExecPlan> getPlanAndFragment(ConnectContext connectContext, String originStmt,
+                                                            boolean supportCreateView) throws Exception {
```
[the remark this answers is not in the document]
Comment (banmoy) [reply fe/fe-core/src/test/java/com/starrocks/utframe/UtFrameUtils.java]: Support `create view view_name as select * from test_binlog [_BINLOG_]` now

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 55 / 59 (93.22%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|3|66.67%|[3565]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|17|82.35%|[2857, 2858, 2859]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToSQLBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/BinlogScanNode.java|16|16|100.00%|[]|

File: be/src/connector/binlog_connector.cpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+BinlogDataSource::BinlogDataSource(const BinlogDataSourceProvider* provider, const TScanRange& scan_range)
+        : _provider(provider), _scan_range(scan_range.binlog_scan_range) {}
+
+Status BinlogDataSource::open(RuntimeState* state) {
+    const TBinlogScanNode& binlog_scan_node = _provider->_binlog_scan_node;
+    _runtime_state = state;
+    _tuple_desc = state->desc_tbl().get_tuple_descriptor(binlog_scan_node.tuple_id);
+    ASSIGN_OR_RETURN(_tablet, _get_tablet())
+    ASSIGN_OR_RETURN(_binlog_read_schema, _build_binlog_schema())
+    LOG(INFO) << "Tablet id " << _tablet->tablet_uid() << ", version " << _scan_range.offset.version << ", seq_id "
```
Comment (mofeiatwork) [inline be/src/connector/binlog_connector.cpp:49]: Make it a vlog?

File: fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewPlanTest.java [hunk 6]
```diff
@@ -65,4 +68,40 @@ public void testCreateIncrementalMV() throws Exception {
...
+                "PLAN FRAGMENT 1\n" +
+                " OUTPUT EXPRS:\n" +
+                "  PARTITION: RANDOM\n" +
+                "\n" +
+                "  STREAM DATA SINK\n" +
+                "    EXCHANGE ID: 01\n" +
+                "    UNPARTITIONED\n" +
+                "\n" +
+                "  0:BinlogScanNode\n" +
+                "     table: binlog_test     tabletList: %s16336,16338"
```
Comment (mofeiatwork) [inline fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewPlanTest.java:103]: Remove the `%` typo?

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 7]
```diff
@@ -3559,8 +3559,11 @@ public ParseNode visitTableAtom(StarRocksParser.TableAtomContext context) {
         TableRelation tableRelation = new TableRelation(tableName, partitionNames, tabletIds);
         if (context.bracketHint() != null) {
             for (Identifier identifier : visit(context.bracketHint().identifier(), Identifier.class)) {
-                if (identifier.getValue().equals("_META_")) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:3562]: Why remove it?

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java [hunk 8]
```diff
@@ -100,12 +107,22 @@ public TableName getResolveTableName() {
         }
     }
 
+    // Maybe throw IllegalArgumentException if hintName is not defined
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java:110]: Throw an explicit exception?

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AstToSQLBuilder.java [hunk 9]
```diff
@@ -215,6 +215,12 @@ public String visitTable(TableRelation node, Void outerScope) {
             StringBuilder sqlBuilder = new StringBuilder();
             sqlBuilder.append(node.getName().toSql());
 
+            for (TableRelation.TableHint hint : node.getTableHints()) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AstToSQLBuilder.java:218]: Wrap an `CollectionUtils.emptyIfNull`?

File: be/src/exec/stream/scan/stream_scan_operator.h [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+#include "exec/pipeline/pipeline_builder.h"
+#include "exec/pipeline/scan/balanced_chunk_buffer.h"
+#include "exec/pipeline/scan/connector_scan_operator.h"
+#include "exec/pipeline/scan/scan_operator.h"
+#include "exec/workgroup/work_group_fwd.h"
+#include "storage/chunk_helper.h"
+
+namespace starrocks {
+
+class ScanNode;
```
Comment (mofeiatwork) [inline be/src/exec/stream/scan/stream_scan_operator.h:28]: I guess you have include the `ScanNode` declaration, so you don't need this forward declaration?

File: be/src/exec/exec_node.cpp [hunk 11]
```diff
@@ -795,6 +813,7 @@ void ExecNode::collect_scan_nodes(vector<ExecNode*>* nodes) {
     collect_nodes(TPlanNodeType::MYSQL_SCAN_NODE, nodes);
     collect_nodes(TPlanNodeType::LAKE_SCAN_NODE, nodes);
     collect_nodes(TPlanNodeType::SCHEMA_SCAN_NODE, nodes);
+    collect_nodes(TPlanNodeType::STREAM_SCAN_NODE, nodes);
```
Comment (mofeiatwork) [inline be/src/exec/exec_node.cpp:816]: it's pretty inefficient to visit the plan tree several times. we should refactor it with a node set

File: be/src/connector/binlog_connector.h [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+    bool stream_data_source() const override { return true; }
+
+protected:
+    ConnectorScanNode* _scan_node;
+    const TBinlogScanNode _binlog_scan_node;
+};
+
+using BinlogMetaFieldMap = std::unordered_map<std::string, VectorizedFieldPtr>;
+const std::string BINLOG_PREFIX = "_binlog";
```
Comment (mofeiatwork) [inline be/src/connector/binlog_connector.h:55]: How about define this constant in thrift, to unify the name in FE and BE

File: be/src/connector/binlog_connector.h [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+class BinlogDataSource final : public DataSource {
+public:
+    ~BinlogDataSource() override = default;
+
+    BinlogDataSource(const BinlogDataSourceProvider* provider, const TScanRange& scan_range);
+    Status open(RuntimeState* state) override;
+    void close(RuntimeState* state) override;
+    Status get_next(RuntimeState* state, ChunkPtr* chunk) override;
+
+    int64_t raw_rows_read() const override;
```
Comment (mofeiatwork) [inline be/src/connector/binlog_connector.h:70]: It's too trivial, these several metrics are used for audit-log and resourcegroup scheduling, IMO only resource-gropu scheduling related metrics are necessary.

File: be/src/connector/binlog_connector.cpp [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    return Status::OK();
+}
+
+StatusOr<TabletSharedPtr> BinlogDataSource::_get_tablet() {
+    TTabletId tablet_id = _scan_range.tablet_id;
+    std::string err;
+    TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, true, &err);
+    if (!tablet) {
+        std::stringstream ss;
+        ss << "failed to get tablet. tablet_id=" << tablet_id << ", reason=" << err;
```
Comment (mofeiatwork) [inline be/src/connector/binlog_connector.cpp:190]: Use `fmt::format` instead of `stringstream`?

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java (hunk 7 above)
Comment (banmoy) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:3565]: Just unify the hint implementation

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java (hunk 8 above)
Comment (banmoy) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/TableRelation.java]: Return a `boolean` for the result

File: fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewPlanTest.java (hunk 6 above)
Comment (banmoy) [reply fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewPlanTest.java]: Done

File: be/src/connector/binlog_connector.cpp (hunk 5 above)
Comment (banmoy) [reply be/src/connector/binlog_connector.cpp]: Done

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AstToSQLBuilder.java (hunk 9 above)
Comment (banmoy) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AstToSQLBuilder.java]: Done

File: be/src/exec/exec_node.cpp (hunk 11 above)
Comment (banmoy) [reply be/src/exec/exec_node.cpp:816]: How about I refactor it in another PR?

File: be/src/connector/binlog_connector.h (hunk 12 above)
Comment (banmoy) [reply be/src/connector/binlog_connector.h]: Done

File: be/src/connector/binlog_connector.cpp (hunk 14 above)
Comment (banmoy) [reply be/src/connector/binlog_connector.cpp:189]: Done

File: be/src/connector/binlog_connector.h (hunk 13 above)
Comment (banmoy) [reply be/src/connector/binlog_connector.h:70]: Maybe I do not get your point. These functions are pure virtual in `DataSource`, and do you mean there is no need to implement them?

**Comment by banmoy** [comment]
@mofeiatwork Resolve most of comments. PTAL again.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 39 / 44 (88.64%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|5|40.00%|[2857, 2858, 2859]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|10|12|83.33%|[117, 118]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToSQLBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/BinlogScanNode.java|13|13|100.00%|[]|

File: be/src/exec/exec_node.cpp (hunk 11 above)
Comment (mofeiatwork) [reply be/src/exec/exec_node.cpp:816]: sure

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by murphyatwork — DISMISSED** [review]

**Review by Astralidea — DISMISSED** [review]

**Comment by banmoy** [comment]
run starrocks_admit_test

**Review by murphyatwork — APPROVED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 39 / 44 (88.64%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|2|5|40.00%|[2857, 2858, 2859]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|10|12|83.33%|[117, 118]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToSQLBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/BinlogScanNode.java|13|13|100.00%|[]|

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**